algebra/group/subgroup.lean

```diff
@@ -0,0 +1,147 @@
+
+@[simp] lemma one_ker_inv {f : G → H} [is_group_hom f] {a b : G} (h : f (a * b⁻¹) = 1) : f a = f b := by rw ←inv_inv (f b); rw [mul f, inv f] at h; exact eq_inv_of_mul_eq_one h
+
+@[simp] lemma inv_ker_one {f : G → H} [is_group_hom f] {a b : G} (h : f a = f b) : f (a * b⁻¹) = 1 :=
+have f a * (f b)⁻¹ = 1, by rw h; apply mul_right_inv,
+by rw [←inv f, ←mul f] at this; exact this
+
+@[simp] lemma ker_inv {f : G → H} [is_group_hom f] {a b : G} (h : a * b⁻¹ ∈ kernel f) : f a = f b := one_ker_inv $ mem_ker_one h
+
+@[simp] lemma inv_ker {f : G → H} [is_group_hom f] {a b : G} (h : f a = f b) : a * b⁻¹ ∈ kernel f := by simp [mem_set_of_eq]; exact inv_ker_one h
+
+lemma one_iff_ker_inv (f : G → H) [is_group_hom f] (a b : G) : f a = f b ↔ f (a * b⁻¹) = 1 :=
+⟨inv_ker_one, one_ker_inv⟩
+
+lemma inv_iff_ker (f : G → H) [w : is_group_hom f] (a b : G) : f a = f b ↔ a * b⁻¹ ∈ kernel f :=
+⟨@inv_ker _ _ _ _ f w _ _, @ker_inv _ _ _ _ f w _ _⟩ -- TODO: I don't understand why I can't just write ⟨inv_ker, ker_inv⟩ here. (This gives typeclass errors; it can't find `w`.)
+
+instance image_subgroup (f : G → H) [is_group_hom f] (S : set G) [subgroup S] : subgroup (f '' S) := {
+  subgroup .
+  mul_mem := assume a₁ a₂ ⟨b₁, hb₁, eq₁⟩ ⟨b₂, hb₂, eq₂⟩,
+  ⟨b₁ * b₂, mul_mem hb₁ hb₂, by simp [eq₁, eq₂, mul f]⟩,
+  one_mem := ⟨1, one_mem S, one f⟩,
+  inv_mem := assume a ⟨b, hb, eq⟩,
+  ⟨b⁻¹, inv_mem hb, by rw inv f; simp *⟩ 
+}
+
+local attribute [simp] subgroup.one_mem 
+                 subgroup.inv_mem 
+                 subgroup.mul_mem
+                 normal_subgroup.normal
+
+instance preimage (f : G → H) [is_group_hom f] (S : set H) [subgroup S] : subgroup (f ⁻¹' S) :=
+by refine {..}; simp [mul f, one f, inv f] {contextual:=tt}
+
+instance preimage_normal (f : G → H) [is_group_hom f] (S : set H) [normal_subgroup S] : normal_subgroup (f ⁻¹' S) :=
+by refine {..}; simp [mul f, one f, inv f] {contextual:=tt}
+
+instance kernel_normal (f : G → H) [is_group_hom f] : normal_subgroup (kernel f) := 
+is_group_hom.preimage_normal f (trivial H)
+
+lemma inj_of_trivial_kernel {f : G → H} [is_group_hom f] (h : kernel f = trivial G) : function.injective f :=
+begin
+  dsimp [function.injective] at *,
+  intros a₁ a₂ hfa,
+  simp [set_eq_def, mem_set_of_eq] at h,
+  have ha : a₁ * a₂⁻¹ = 1, by rw ←h; exact inv_ker_one hfa,
+  rw [eq_inv_of_mul_eq_one ha, inv_inv a₂]
+end
+
+lemma trivial_kernel_of_inj {f : G → H} [is_group_hom f] (h : function.injective f) : kernel f = trivial G :=
+begin
+  dsimp [function.injective] at *,
+  simp [set_eq_def, mem_set_of_eq],
+  intro,
+  split,
+  { assume hx, 
+    have hi : f x = f 1 := by simp [one f, hx],
+    simp [h hi, one f] },
+  { assume hx, simp [hx, one f] }
+end
+
+lemma inj_iff_trivial_kernel {f : G → H} [w : is_group_hom f] : function.injective f ↔ kernel f = trivial G :=
+⟨@trivial_kernel_of_inj _ _ _ _ f w, @inj_of_trivial_kernel _ _ _ _ f w⟩ -- TODO again, why can't it find w by itself?
```

Comment:
Make `f` explicit in this and the last two lemmas, and that should fix your typeclass problem. You may be getting a sense for why I thought making `is_group_hom` a typeclass doesn't gain you that much, except possibly chains like `is_group_hom.comp`.

Reply:
Unfortunately it doesn't fix it. Commit coming in a moment.